### ISSUE TITLE
Questions about training hyperparameters

### ISSUE THREAD

**Issue by BlankCheng** [facebookresearch/DPR#75]
Hello! Many thanks for your organized code, it's a great help for me!
I have some questions when training retriever and reader with provided dataset from `download_data.sh`. A single server with 8*24G Titan RTX is used.

1. `batchsize=16` is suggested for `train_dense_encoder.py`, but it seems only 35G gpu memory are used. Is there any tradeoff not to set a larger batchsize since 8*32G V100 are used?

2. In README, it is said ` train_reader.py` achieves best performance after 16-18 epochs, or ~60k steps. But if `batchsize=16`, one epoch contains only ~500 steps. "16-18 epochs" may not be equivalent to "60k steps". 

3. And when I train the reader, it seems validation EM reaches 27% at best,  after training 40 epochs, much lower than 41.5% in paper. I use 'train_0.pkl" as train_file, but there are other seven "train_*.pkl". Should I use all of them? If so, used separately for 8 rounds or together in one train? My training time of reader is only a quarter of that of retriever.  So I think I must have done it by mistake.
The script I used to train reader:
```
 python train_reader.py --seed 42 --learning_rate 1e-5 --eval_step 2000 --do_lower_case \
--eval_top_docs 50 --encoder_model_type hf_bert --pretrained_model_cfg bert-base-uncased \
--train_file ./dpr/data/reader/nq/single/train_0.pkl --dev_file ./dpr/data/reader/nq/single/dev.pkl \
 --output_dir ./dpr/results --warmup_steps 0 --sequence_length 350 --batch_size 16 --passages_per_question 24 \
--num_train_epochs 100000 --dev_batch_size 72 --passages_per_question_predict 50
```

Best
Cheng

**Comment by vlad-karpukhin** [comment]
Hello,

1. we tried 256 total effective batch size but the performance was ~ same as 128 total (16x8). We didn't play much with adjusting the learning rate based on the batch size though. It would interesting to know if it is possible to get better results with larger batch size.

2. "but there are other seven "train_*.pkl"." - this is because preprocessing is done in multithreaded way and each .pkl is the output from a single thread. If you use just one, you effectively use 1/8 of all rain data. You should use 
--train_file ./dpr/data/reader/nq/single/train_*.pkl

**Comment by BlankCheng** [comment]
Got it, thank you! I misunderstand `batchsize=16` in train retriever since multiple nodes are used and actually `batchsize=8*16`, while I use a single server. And I‘ll try training the reader again.
I'll close it.
